### ISSUE TITLE
Rule Request: Prefer `Self` over `type(of: self)`

### ISSUE THREAD

**Issue by marcelofabri** [realm/SwiftLint#3003]
### New Issue Checklist

- [x] Updated SwiftLint to the latest version
- [x] I searched for [existing GitHub issues](https://github.com/realm/SwiftLint/issues)

https://github.com/realm/SwiftLint/issues/2944 and https://github.com/realm/SwiftLint/issues/2921 are related, but I think this should be a separate rule.

### New rule request

1. Why should this rule be added? Share links to existing discussion about what
   the community thinks about this.

[Swift 5.1](https://mjtsai.com/blog/2019/09/23/swift-5-1-released/#comment-3071047) introduces the [`Self` type](https://docs.swift.org/swift-book/ReferenceManual/Types.html#grammar_self-type), which can be used in place of `type(of: self)`.

2. Provide several examples of what _would_ and _wouldn't_ trigger violations.

```swift
// ok
class A {
  func foo() {
    Self.bar() 
   }
}

class A {
  func foo(param: B) {
    type(of: param).bar() 
   }
}

// triggers
class A {
  func foo() {
    type(of: self).bar() 
   }
}
```
3. Should the rule be configurable, if so what parameters should be configurable?

Just severity.

4. Should the rule be opt-in or enabled by default? Why?
   See [README.md](../README.md#opt-in-rules) for guidelines on when to mark a rule as opt-in.

Opt-in.

**Comment by jpsim** [comment]
Why opt-in? Are there cases where this isn't preferred or where the semantics are different?

**Comment by jpsim** [comment]
> Why opt-in?

Ah I know now, it's for support for Swift < 5.1

**Comment by marcelofabri** [comment]
Also I’m not sure if the community has settled on one or another

**Comment by cltnschlosser** [comment]
Is there a performance benefit to `Self`? or does the runtime still need to look up the type.

**Comment by jpsim** [comment]
Thanks for the thoughtful analysis, @cltnschlosser ! ❤️
